### PR TITLE
Feature flag design

### PR DESCRIPTION
**Pull request by nrb**
* [ ] Address graduation & removal of feature flags. Do the flags always stay but the conditionals get removed?
* [ ] Add command for listing known features?

Signed-off-by: Nolan Brubaker <brubakern@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
Approach seems reasonable. We'll want to add a docs page listing the valid features and whether they're enabled/disabled by default.

Just looking at Kubernetes feature gates (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) - should we make the flag take a series of key-value pairs, e.g. `CSISupport=true,Replication=false,...`?  This would make it easier to turn off a feature that's on by default, for example.

**Comment by nrb** [comment]
That's true. I was more thinking that we only put feature flags for things that were in-progress, and when they graduated those flags would be removed.

That lifecycle of a feature and its associated flag(s) should be added to this.

**Comment by prydonius** [comment]
> I was more thinking that we only put feature flags for things that were in-progress, and when they graduated those flags would be removed.

I would agree, not sure that we will really need to keep these feature flags once we want to make them default. I suppose this is useful in Kubernetes due to some features depending on specific environments and requirements.

**Comment by skriss** [comment]
Alright, that works for me. We can always change later if needed.

File: design/feature-flags.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+A new command line flag, `--features` will be added to the root `velero` command.
+
+`--features` will accept a comma-separated list of features, such as `--features EnableCSI,Replication`.
+Each feature listed will correspond to a key in a map in `pkg/features/features.go` defining whether a feature should be enabled.
+
+Any code implementing the feature would then import the map and look up the key's value.
+
+For the Velero client, a `features` key can be added to the config.json file for more convenient client invocatons.
+
+## Detailed Design
```
Comment (prydonius) [inline design/feature-flags.md]: We might want to drop all the placeholder stuff after this, so it's less confusing to read this proposal

File: design/feature-flags.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+For the Velero client, a `features` key can be added to the `config.json` file for more convenient client invocations.
+
+## Detailed Design
+
+A new `features` package will be introduced with these basic structs:
+
+```go
+type FeatureFlagSet struct {
+    flags map[string]bool
```
Comment (prydonius) [inline design/feature-flags.md]: Guess this could also be a `sets.String` type, unless we want to keep the option of being able to disable a flag?

Comment (nrb) [reply design/feature-flags.md]: Oh, yeah, that's a good idea.

**Comment by skriss** [comment]
ready to merge when it moves out of draft and changes to `Accepted`.

**Comment by nrb** [comment]
@skriss updated status and squashed

**Pull request merged**